src/ui/api/projectmember.go

```diff
@@ -0,0 +1,216 @@
+
+	if !(pma.Ctx.Input.IsGet() && pma.SecurityCtx.HasReadPerm(pid) ||
+		pma.SecurityCtx.HasAllPerm(pid)) {
+		pma.HandleForbidden(pma.SecurityCtx.GetUsername())
+		return
+	}
+
+	pmid, err := pma.GetInt64FromPath(":pmid")
+	if err != nil {
+		log.Errorf("Failed to get pmid from path, error %v", err)
+	}
+	if pmid > 0 {
+		pma.id = int(pmid)
+	}
+}
+
+//Get ...
+func (pma *ProjectMemberAPI) Get() {
+	projectID := pma.project.ProjectID
+	queryMember := models.Member{}
+	queryMember.ProjectID = projectID
+	pma.Data["json"] = make([]models.Member, 0)
+	if pma.id == 0 {
+		//member id not set, return all member of current project
+		memberList, err := project.GetProjectMember(queryMember)
+		if err != nil {
+			log.Errorf("Failed to query database for member list, error: %v", err)
+			pma.RenderError(http.StatusInternalServerError, "Internal Server Error")
+			return
+		}
+		if len(memberList) > 0 {
+			pma.Data["json"] = memberList
+		}
+	} else {
+		//return a specific member
+		queryMember.ID = pma.id
+		memberList, err := project.GetProjectMember(queryMember)
+		if err != nil {
+			log.Errorf("Failed to query database for member list, error: %v", err)
+			pma.RenderError(http.StatusInternalServerError, "Internal Server Error")
+			return
+		}
+		if len(memberList) > 0 {
+			pma.Data["json"] = memberList[0]
+		}
+	}
+	pma.ServeJSON()
+}
+
+// Post ... Add a project member
+func (pma *ProjectMemberAPI) Post() {
+	projectID := pma.project.ProjectID
+	var request models.MemberReq
+	pma.DecodeJSONReq(&request)
+	err := AddOrUpdateProjectMember(projectID, request)
+	if err != nil {
+		log.Errorf("Failed to add project member, error: %v", err)
+		pma.RenderError(http.StatusInternalServerError, "Failed to add project member")
+		return
+	}
+}
+
+// Put ... Update an exist project member
+func (pma *ProjectMemberAPI) Put() {
+	pid := pma.project.ProjectID
+	pmID := pma.id
+	if pmID == 0 {
+		log.Errorf("Failed to update DB to add project user role, project id: %d, pmid : %d", pid, pmID)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+	var req models.Member
+	pma.DecodeJSONReq(&req)
+	err := project.UpdateProjectMemberRole(pmID, req.Role)
+	if err != nil {
+		log.Errorf("Failed to update DB to add project user role, project id: %d, pmid : %d, role id: %d", pid, pmID, req.Role)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+}
+
+// Delete ...
+func (pma *ProjectMemberAPI) Delete() {
+	pmid := pma.id
+	err := project.DeleteProjectMemberByID(pmid)
+	if err != nil {
+		log.Errorf("Failed to delete project roles for user, project member id: %d, error: %v", pmid, err)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+}
+
+// AddOrUpdateProjectMember ... If the project member relationship does not exist, create it. if exist, update it
+func AddOrUpdateProjectMember(projectID int64, request models.MemberReq) error {
```

Comment:
IMO, this should be simplified, and the logic should be in API layer.

Reply:
It is used by src/ui/auth/ldap_test.go

Reply:
Referenced by ui/auth/ldap_test.go